core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -88,6 +88,8 @@ public class AlluxioMasterProcess extends MasterProcess {
   private final MasterUfsManager mUfsManager = new MasterUfsManager();
 
   private AlluxioExecutorService mRPCExecutor = null;
+  /** See {@link #isStopped()}. */
+  protected boolean mIsStopped = false;
```

Comment:
there were problems with different threads calling the shutdown hooks right? which means this could be called concurrently, is that still possible?

If so you probably want to use an atomic boolean here and use compare and set to ensure to get only one process successfully calling close

Reply:
I used `synchronized` blocks instead as I only want `mIsStopped = true` if `stop()` finished normally.

Reply:
Does that mean the assumption is that if one call failed to stop the process and threw an exception, then there may be another call to stop later that is able to stop the process correctly? I think these calls originally were not designed to be thread safe, but have been used from different threads so their behavior is sort of unclear. Maybe you don't want to fix all the issues in this PR, but could maybe make a note on the expected behavior.

Reply:
I guess yes, the assumption is that a first call to `stop` might fail while a second call to `stop` might succeed. The most important part is that only one thread may call `stop` at a time and `mIsStopped` is only true if all the resources have been properly cleaned up. 
I also think that multiple concurrent calls to stop might release resources in unexpected and undefined orders, which could make releasing the resources impossible. I believe that synchronizing the release of resources is positive.

Reply:
Ok could you make a comment in the code mentioning this please, and that it partially addresses the issue AE-2488